File: src-tauri/src/wiki/frontmatter.rs
// Wiki frontmatter 解析 + wikilink 提取
//
// 设计要点:
// - 极简 YAML 子集解析:只支持 string / [a, b, c] / 多行
// - 不引 serde_yaml(减少依赖)
// - 字段集合:title, type, tags, related, sources, created, updated
// - wikilink 提取支持 [[X]] 和 [[X|Y]]

use std::collections::HashMap;

// ============================================================================
// 类型
// ============================================================================

/// 解析后的 frontmatter
#[derive(Debug, Clone, PartialEq, Default)]
pub struct Frontmatter {
    pub title: Option<String>,
    pub page_type: Option<String>,
    pub tags: Vec<String>,
    pub related: Vec<String>,
    pub sources: Vec<String>,
    pub created: Option<String>,
    pub updated: Option<String>,
    /// 未知字段(透传保留)
    pub extra: HashMap<String, String>,
}

/// 解析结果:(frontmatter, body)
#[derive(Debug, Clone, PartialEq)]
pub struct ParsedDoc {
    pub frontmatter: Frontmatter,
    pub body: String,
}

// ============================================================================
// 解析
// ============================================================================

/// 解析 `---\n...\n---\nbody` 格式
pub fn parse(content: &str) -> ParsedDoc {
    // 找前导 `---` 起始
    if !content.starts_with("---\n") && content != "---" {
        return ParsedDoc {
            frontmatter: Frontmatter::default(),
            body: content.to_string(),
        };
    }
    // 找第二个 `---` 结束
    let after_first = if content == "---" { "" } else { &content[4..] };
    let close = after_first.find("\n---");
    let (yaml, body) = match close {
        Some(pos) => {
            let yaml = &after_first[..pos];
            // body 跳过 `---` 和后续最多 2 个换行(对应 `---\n\n` 模式)
            let rest = &after_first[pos + 4..];
            let body = rest
                .strip_prefix('\n')
                .unwrap_or(rest)
                .strip_prefix('\n')
                .unwrap_or(rest.strip_prefix('\n').unwrap_or(rest))
                .to_string();
            (yaml, body)
        }
        None => return ParsedDoc {
            frontmatter: Frontmatter::default(),
            body: content.to_string(),
        },
    };
    let frontmatter = parse_yaml_subset(yaml);
    ParsedDoc { frontmatter, body }
}

fn parse_yaml_subset(yaml: &str) -> Frontmatter {
    let mut fm = Frontmatter::default();
    for line in yaml.lines() {
        let line = line.trim_end();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        let Some((key, value)) = line.split_once(':') else {
            continue;
        };
        let key = key.trim();
        let value = value.trim();
        match key {
            "title" => fm.title = Some(value.to_string()),
            "type" => fm.page_type = Some(value.to_string()),
            "tags" => fm.tags = parse_list_value(value),
            "related" => fm.related = parse_related_value(value),
            "sources" => fm.sources = parse_list_value(value),
            "created" => fm.created = Some(value.to_string()),
            "updated" => fm.updated = Some(value.to_string()),
            _ => {
                fm.extra.insert(key.to_string(), value.to_string());
            }
        }
    }
    fm
}

/// 解析 `[a, b, c]` 或 `a, b, c` 格式
fn parse_list_value(value: &str) -> Vec<String> {
    let trimmed = value.trim();
    let inner = if trimmed.starts_with('[') && trimmed.ends_with(']') {
        &trimmed[1..trimmed.len() - 1]
    } else {
        trimmed
    };
    inner
        .split(',')
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
        .collect()
}

/// related 字段特殊处理:每项是 `[[X]]` 或 `[[X|Y]]`,提取 X
fn parse_related_value(value: &str) -> Vec<String> {
    extract_wikilinks(value)
}

// ============================================================================
// 序列化
// ============================================================================

/// 序列化为 `---\n...\n---\nbody` 格式
pub fn serialize(doc: &ParsedDoc) -> String {
    let mut out = String::from("---\n");
    if let Some(t) = &doc.frontmatter.title {
        out.push_str(&format!("title: {}\n", t));
    }
    if let Some(t) = &doc.frontmatter.page_type {
        out.push_str(&format!("type: {}\n", t));
    }
    if !doc.frontmatter.tags.is_empty() {
        out.push_str(&format!("tags: [{}]\n", doc.frontmatter.tags.join(", ")));
    }
    if !doc.frontmatter.related.is_empty() {
        let links: Vec<String> = doc.frontmatter.related.iter().map(|r| format!("[[{}]]", r)).collect();
        out.push_str(&format!("related: {}\n", links.join(" ")));
    }
    if !doc.frontmatter.sources.is_empty() {
        out.push_str(&format!("sources: [{}]\n", doc.frontmatter.sources.join(", ")));
    }
    if let Some(c) = &doc.frontmatter.created {
        out.push_str(&format!("created: {}\n", c));
    }
    if let Some(u) = &doc.frontmatter.updated {
        out.push_str(&format!("updated: {}\n", u));
    }
    for (k, v) in &doc.frontmatter.extra {
        out.push_str(&format!("{}: {}\n", k, v));
    }
    out.push_str("---\n");
    if !doc.body.is_empty() {
        out.push_str(&doc.body);
        if !doc.body.ends_with('\n') {
            out.push('\n');
        }
    }
    out
}

// ============================================================================
// wikilink 提取
// ============================================================================

/// 从内容中提取所有 `[[X]]` 或 `[[X|Y]]`,返回 X 列表(顺序,去重)
pub fn extract_wikilinks(content: &str) -> Vec<String> {
    let mut out = Vec::new();
    let bytes = content.as_bytes();
    let mut i = 0;
    while i + 1 < bytes.len() {
        if bytes[i] == b'[' && bytes[i + 1] == b'[' {
            // 找 `]]`
            let mut j = i + 2;
            while j + 1 < bytes.len() && !(bytes[j] == b']' && bytes[j + 1] == b']') {
                j += 1;
            }
            if j + 1 < bytes.len() {
                let inner = &content[i + 2..j];
                // 可能有 `|` 分隔符
                let target = inner.split('|').next().unwrap_or(inner).trim();
                if !target.is_empty() && !out.contains(&target.to_string()) {
                    out.push(target.to_string());
                }
                i = j + 2;
                continue;
            }
        }
        i += 1;
    }
    out
}

// ============================================================================
// Tests
// ============================================================================

#[cfg(test)]
mod tests {
    use super::*;

    // ---- parse ----

    #[test]
    fn parse_no_frontmatter_returns_empty() {
        let content = "# Hello\n\nBody text.";
        let p = parse(content);
        assert_eq!(p.frontmatter, Frontmatter::default());
        assert_eq!(p.body, content);
    }

    #[test]
    fn parse_simple_frontmatter() {
        let content = "---\ntitle: Albert Einstein\ntype: entity\n---\n\n# Body";
        let p = parse(content);
        assert_eq!(p.frontmatter.title.as_deref(), Some("Albert Einstein"));
        assert_eq!(p.frontmatter.page_type.as_deref(), Some("entity"));
        assert_eq!(p.body, "# Body");
    }

    #[test]
    fn parse_list_field_brackets() {
        let content = "---\ntags: [physicist, nobel-prize]\n---\n";
        let p = parse(content);
        assert_eq!(p.frontmatter.tags, vec!["physicist", "nobel-prize"]);
    }

    #[test]
    fn parse_list_field_csv() {
        let content = "---\ntags: a, b, c\n---\n";
        let p = parse(content);
        assert_eq!(p.frontmatter.tags, vec!["a", "b", "c"]);
    }

    #[test]
    fn parse_related_field_with_wikilinks() {
        let content = "---\nrelated: [[Theory of Relativity]] [[Nobel Prize]]\n---\n";
        let p = parse(content);
        assert_eq!(p.frontmatter.related, vec!["Theory of Relativity", "Nobel Prize"]);
    }

    #[test]
    fn parse_related_with_alias() {
        let content = "---\nrelated: [[X|Y]]\n---\n";
        let p = parse(content);
        assert_eq!(p.frontmatter.related, vec!["X"]);
    }

    #[test]
    fn parse_sources_field() {
        let content = "---\nsources: [raw/sources/a.pdf, raw/sources/b.md]\n---\n";
        let p = parse(content);
        assert_eq!(p.frontmatter.sources, vec!["raw/sources/a.pdf", "raw/sources/b.md"]);
    }

    #[test]
    fn parse_dates_field() {
        let content = "---\ncreated: 2026-06-12\nupdated: 2026-06-13\n---\n";
        let p = parse(content);
        assert_eq!(p.frontmatter.created.as_deref(), Some("2026-06-12"));
        assert_eq!(p.frontmatter.updated.as_deref(), Some("2026-06-13"));
    }

    #[test]
    fn parse_unknown_fields_go_to_extra() {
        let content = "---\ncustom_field: hello\n---\n";
        let p = parse(content);
        assert_eq!(p.frontmatter.extra.get("custom_field").map(String::as_str), Some("hello"));
    }

    // ---- serialize ----

    #[test]
    fn serialize_roundtrip() {
        let original = "---\ntitle: X\ntype: source\ntags: [a, b]\n---\n\nbody text";
        let p = parse(original);
        let s = serialize(&p);
        let p2 = parse(&s);
        assert_eq!(p.frontmatter, p2.frontmatter);
    }

    #[test]
    fn serialize_empty_frontmatter_still_has_delimiters() {
        let p = ParsedDoc {
            frontmatter: Frontmatter::default(),
            body: "body".to_string(),
        };
        let s = serialize(&p);
        assert!(s.starts_with("---\n"));
        assert!(s.contains("\n---\n"));
    }

    // ---- extract_wikilinks ----

    #[test]
    fn extract_simple_wikilink() {
        let s = "see [[Theory of Relativity]] for details";
        assert_eq!(extract_wikilinks(s), vec!["Theory of Relativity"]);
    }

    #[test]
    fn extract_multiple_wikilinks() {
        let s = "links: [[A]] and [[B]] and [[C]]";
        assert_eq!(extract_wikilinks(s), vec!["A", "B", "C"]);
    }

    #[test]
    fn extract_wikilink_with_alias() {
        let s = "see [[Page Name|display text]]";
        assert_eq!(extract_wikilinks(s), vec!["Page Name"]);
    }

    #[test]
    fn extract_dedupes_repeated_links() {
        let s = "[[A]] [[B]] [[A]]";
        assert_eq!(extract_wikilinks(s), vec!["A", "B"]);
    }

    #[test]
    fn extract_no_wikilinks() {
        assert!(extract_wikilinks("plain text").is_empty());
    }

    #[test]
    fn extract_unclosed_bracket_ignored() {
        // `[[X` 没有 `]]` 结尾应忽略
        assert!(extract_wikilinks("broken [[X").is_empty());
    }
}

File: backend/wiki/frontmatter.py
"""YAML frontmatter 解析和序列化。

支持解析 Markdown 文件中的 YAML frontmatter，提取 wikilinks，序列化回 Markdown。
用于 Wiki 页面（区别于 backend/skills/skill_md/frontmatter.py 用于 SKILL.md）。
"""

from dataclasses import dataclass, field


@dataclass
class Frontmatter:
    """YAML frontmatter 数据。"""

    title: str | None = None
    page_type: str | None = None  # "source"/"entity"/"concept"/...
    tags: list[str] = field(default_factory=list)
    related: list[str] = field(default_factory=list)  # wikilink 目标
    sources: list[str] = field(default_factory=list)
    created: str | None = None
    updated: str | None = None
    extra: dict[str, str] = field(default_factory=dict)  # 未知字段


@dataclass
class ParsedDoc:
    """解析后的文档。"""

    frontmatter: Frontmatter
    body: str


def parse(content: str) -> ParsedDoc:
    """解析 Markdown 内容，提取 frontmatter 和 body。

    Args:
        content: Markdown 文件内容

    Returns:
        ParsedDoc: 包含 frontmatter 和 body 的文档
    """
    if not content.startswith("---\n"):
        return ParsedDoc(Frontmatter(), content)

    # 查找结束的 ---
    remainder = content[4:]
    end_idx = remainder.find("\n---")
    if end_idx == -1:
        return ParsedDoc(Frontmatter(), content)

    yaml_str = remainder[:end_idx]
    body_start = end_idx + 5  # 跳过 "\n---"
    # 跳过最多 2 个换行
    while body_start < len(remainder) and remainder[body_start] == "\n":
        body_start += 1
    body = remainder[body_start:]

    # 解析 YAML（简化版，逐行解析）
    frontmatter = _parse_yaml(yaml_str)
    return ParsedDoc(frontmatter, body)


def _parse_yaml(yaml_str: str) -> Frontmatter:
    """解析 YAML 字符串为 Frontmatter。

    Args:
        yaml_str: YAML 字符串

    Returns:
        Frontmatter: 解析后的 frontmatter
    """
    fm = Frontmatter()

    for raw_line in yaml_str.split("\n"):
        line = raw_line.strip()
        if not line or line.startswith("#"):
            continue

        # 分割 key: value
        colon_idx = line.find(":")
        if colon_idx == -1:
            continue

        key = line[:colon_idx].strip()
        value = line[colon_idx + 1 :].strip()

        # 映射已知字段
        if key == "title":
            fm.title = value
        elif key == "type":
            fm.page_type = value
        elif key == "tags":
            fm.tags = _parse_list_value(value)
        elif key == "related":
            fm.related = _parse_related_value(value)
        elif key == "sources":
            fm.sources = _parse_list_value(value)
        elif key == "created":
            fm.created = value
        elif key == "updated":
            fm.updated = value
        else:
            fm.extra[key] = value

    return fm


def _parse_list_value(value: str) -> list[str]:
    """解析列表值（支持 [a, b, c] 格式）。

    Args:
        value: 值字符串

    Returns:
        list[str]: 解析后的列表
    """
    if value.startswith("[") and value.endswith("]"):
        value = value[1:-1]

    items = [item.strip() for item in value.split(",")]
    return [item for item in items if item]


def _parse_related_value(value: str) -> list[str]:
    """解析 related 字段（提取 [[X]] wikilinks）。

    Args:
        value: 值字符串

    Returns:
        list[str]: 提取的 wikilinks
    """
    return extract_wikilinks(value)


def extract_wikilinks(content: str) -> list[str]:
    """提取内容中的所有 wikilinks [[X]]。

    Args:
        content: Markdown 内容

    Returns:
        list[str]: 提取的 wikilinks（去重，保持顺序）
    """
    links = []
    seen = set()

    i = 0
    while i < len(content) - 1:
        if content[i] == "[" and content[i + 1] == "[":
            # 找到 ]]
            end_idx = content.find("]]", i + 2)
            if end_idx != -1:
                inner = content[i + 2 : end_idx]
                # 处理 [[X|Y]] 格式，取 X
                if "|" in inner:
                    inner = inner.split("|")[0]
                inner = inner.strip()
                if inner and inner not in seen:
                    links.append(inner)
                    seen.add(inner)
                i = end_idx + 2
                continue
        i += 1

    return links


def serialize(doc: ParsedDoc) -> str:
    """序列化 ParsedDoc 为 Markdown 字符串。

    Args:
        doc: 解析后的文档

    Returns:
        str: Markdown 字符串
    """
    lines = ["---"]

    if doc.frontmatter.title:
        lines.append(f"title: {doc.frontmatter.title}")
    if doc.frontmatter.page_type:
        lines.append(f"type: {doc.frontmatter.page_type}")
    if doc.frontmatter.tags:
        lines.append(f"tags: [{', '.join(doc.frontmatter.tags)}]")
    if doc.frontmatter.related:
        related_str = " ".join(f"[[{r}]]" for r in doc.frontmatter.related)
        lines.append(f"related: {related_str}")
    if doc.frontmatter.sources:
        lines.append(f"sources: [{', '.join(doc.frontmatter.sources)}]")
    if doc.frontmatter.created:
        lines.append(f"created: {doc.frontmatter.created}")
    if doc.frontmatter.updated:
        lines.append(f"updated: {doc.frontmatter.updated}")

    # 额外字段
    for key, value in doc.frontmatter.extra.items():
        lines.append(f"{key}: {value}")

    lines.append("---")
    lines.append(doc.body)

    return "\n".join(lines)
